캠페인 생성 API 입력 유효성 검증 로직 구현.

Rebwon opened the pull request:
Resolves #{[2](https://github.com/Self-Develop-Side/ad-system/issues/2)}

## 해결하려는 문제가 무엇인가요?
- .

## 어떻게 해결했나요?
- .

## 어떤 부분에 집중하여 리뷰해야 할까요?
- .

## 참고 자료
- .

## RCA 룰
r: 꼭 반영해 주세요. 적극적으로 고려해 주세요. (Request changes)
c: 웬만하면 반영해 주세요. (Comment)
a: 반영해도 좋고 넘어가도 좋습니다. 그냥 사소한 의견입니다. (Approve)

solomon-maeng committed:
캠페인 생성 시 입력 값이 null인 경우를 검증하는 테스트 추가

solomon-maeng committed:
입력 값이 NULL인 경우에 대한 테스트를 성공하도록 코드 추가

- Kotlin + Spring을 사용할 때 RequestBody에 사용하는 개체의 변수 선언이 val(value)로 되어 있을 때, NULL 입력을 클라이언트로부터 받게 되면 HttpMessageNotReadableException이 발생합니다.
- 이전에 작성했던 테스트 케이스에서 HttpMessageNotReadableException이 발생하는 부분을 해결하고, RFC 7807 문서에 기록된 에러 응답 표준 포맷을 지키도록 Spring에서 제공하는 ProblemDetail 개체를 사용해 클라이언트에게 에러 응답을 제공하도록 코드를 추가했습니다.

solomon-maeng committed:
ktlint에서 trailling comma를 허용하도록 코드 수정

solomon-maeng committed:
캠페인 생성 시 입력 값이 공백인 경우를 검증하는 테스트 추가

solomon-maeng committed:
WildCard import와 Multiline Annotation 룰을 비활성화

wenodev commented on src/main/kotlin/com/example/ad/contract/CreateCampaignApi.kt:
```diff
@@ -0,0 +1,47 @@
+            .body(ProblemDetails.forNullInput())
+    }
+
+    companion object {
+        const val PATH = "/api/v1/campaigns/contract"
+    }
```
a: `PATH` 라는 변수는 테스트코드에서만 사용되는데요.
테스트코드에서 선언하는게 좋지 않나요?

Rebwon replied:
사실 좀 고민이 됬던 포인트긴 한데요. PATH 관리를 프로덕션 기준으로 `@PostMapping(PATH)`로 쓸까 싶었는데, 한번 잘 정의하면 거의 안 바뀔 확률이 높다보니 말씀하신 것처럼 테스트 코드에서만 선언하는 것이 좋아보입니다.

wenodev commented on src/main/kotlin/com/example/ad/contract/CreateCampaignApi.kt:
```diff
@@ -36,12 +46,27 @@ object ProblemDetails {
 }
 
 data class CreateCampaignRequest(
+    @field:NotBlank
```
a : `@field` 이 부분을 명시적으로 하면 어떤 장점이 있는지 궁금합니다.
필드에 선언되었기 때문에 불필요하다고도 느껴져서요

Rebwon replied:
Kotlin에서 `@NotNull`, `@NotBlank`와 같은 Validation 애너테이션을 사용하려면, `@field:{AnnotationName}` 라는 형식으로 써줘야 합니다.

intellij에서 `@field:NotBlank`를 `@NotBlank`로 바꾼 뒤에 코드를 다시 컴파일 하시고, intellij 도구 메뉴 -> Kotlin -> Kotlin 바이트코드 표시 -> 디컴파일을 클릭해보시면 컴파일 된 코드가 다르게 나오는 걸 확인하실 수 있습니다.

## @field:NotBlank의 경우
<img width="311" alt="스크린샷 2023-10-07 오전 10 15 40" src="https://github.com/Self-Develop-Side/ad-system/assets/31135322/d9a9e41d-c279-4f32-b915-bd9aca24ef29">

디컴파일 된 코드
<img width="454" alt="스크린샷 2023-10-07 오전 10 15 45" src="https://github.com/Self-Develop-Side/ad-system/assets/31135322/516077a5-d13f-4bd0-88de-3a1b59d06049">

## @NotBlank만 했을 경우
<img width="306" alt="스크린샷 2023-10-07 오전 10 16 10" src="https://github.com/Self-Develop-Side/ad-system/assets/31135322/adb11e65-3de4-481d-8ea1-3ccaa8d7273b">

디컴파일 된 코드
<img width="549" alt="스크린샷 2023-10-07 오전 10 16 17" src="https://github.com/Self-Develop-Side/ad-system/assets/31135322/02c5bc74-1f9f-441e-9240-bfaa0bc25881">

wenodev replied:
선택이 아니라 필수군요. 감사합니다. 👍

solomon-maeng committed:
API Path 상수를 테스트 코드로 이동

solomon-maeng committed:
입력 값이 공백인 테스트를 입력 값이 유효한 지 검증하도록 수정

- 유효성 검증 시 MethodArgumentNotValidException의 BindingResult 내부에 존재하는 값을 ValidationError 개체에 담아 ProblemDetail로 반환하도록 코드를 수정했습
니다.
- Json Body 검증을 효과적으로 할 수 있도록 JsonUnit 의존성을 추가했습니다.
- CreateCampaignRequest 개체의 검증 애너테이션이 추가될 때 유효성 검증 테스트에서 검증할 수 있도록 테스트 코드를 수정했습니다.

solomon-maeng committed:
CI에서 테스트가 실패하는 이슈 수정

- 로컬 환경 Locale 설정과 CI 환경 Locale 설정이 서로 달라 테스트 코드가 실패하는 이슈를 Locale 설정을 추가하여 해결.

dding94 commented on src/main/kotlin/com/example/ad/contract/CreateCampaignApi.kt:
```diff
@@ -0,0 +1,82 @@
+data class CreateCampaignRequest(
+    @field:NotBlank
+    val clientId: String,
+    @field:NotBlank
+    @field:Size(min = 25, max = 50)
+    val name: String,
+    @field:NotBlank
+    @field:Size(min = 5, max = 10)
+    val createdBy: String,
```
아직 도메인이 잘 이해가 안되어 질문 드립니다. 두 필드 사이즈를 이렇게 정의하신 이유가 궁금합니다.

Rebwon replied:
name 값과 createdBy 값을 DB에서 VARCHAR 사이즈로 제한하기 전에 클라이언트 요청에서 제한하고 싶었구요. 

적절한 사이즈 값은 실제 데이터를 만들고 통합 테스트를 진행하면서 수정해야 하므로 임의로 설정했습니다.

dding94 replied:
일단 클라이언트 요청에서 제한을 임의의 사이즈로 하고 실제 나중에 변경이 되는거군요. 감사합니다 👍
